File: leetcode/python/71_simplify_path.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        dirs = path.split('/')
        stack: list[str] = []
        for d in dirs:
            if d == '..':
                if stack:
                    stack.pop()
            elif not d or d == '.':
                continue
            else:
                stack.append(d)
        return '/' + '/'.join(stack)


s = Solution()
print(s.simplifyPath(path = "/home/"))
print(s.simplifyPath(path = "/../"))
print(s.simplifyPath(path = "/home//foo/"))

File: leetcode/c++/71_simplify_path.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string simplifyPath(string path) {
        vector<string> stack;
        int i = 0;
        while (i < path.size()) {
            while (i < path.size() && path[i] == '/') {
                i++;
            }
            if (i == path.size()) {
                break;
            }
            int start = i;
            while (i < path.size() && path[i] != '/') {
                i++;
            }
            int end = i - 1;
            string sub = path.substr(start, end - start + 1);
            if (sub == "..") {
                if (!stack.empty()) {
                    stack.pop_back();
                }
            } else if (sub != ".") {
                stack.push_back(sub);
            }
        }
        if (stack.empty()) {
            return "/";
        }
        string res;
        for (auto s : stack) {
            res += "/" + s;
        }
        return res;
    }
};